ipie/addons/thermal/estimators/handler.py

```diff
@@ -0,0 +1,251 @@
+                fh5["block_size_1/walker_prop_header"] = self.walker_header
+                for k, o in self.items():
+                    fh5[f"block_size_1/shape/{k}"] = o.shape
+                    fh5[f"block_size_1/size/{k}"] = o.size
+                    fh5[f"block_size_1/scalar/{k}"] = int(o.scalar_estimator)
+                    fh5[f"block_size_1/names/{k}"] = " ".join(name for name in o.names)
+                    fh5[f"block_size_1/offset/{k}"] = self.num_walker_props + self.get_offset(k)
+        if comm.rank == 0:
+            print(header)
+
+    def dump_metadata(self):
+        with h5py.File(self.filename, "a") as fh5:
+            fh5["metadata"] = self.json_string
+
+    def increment_file_number(self):
+        self.index = self.index + 1
+        self.filename = self.basename + f".{self.index}.h5"
+
+    def compute_estimators(self, hamiltonian, trial, walker_batch):
+        """Update estimators with bached psi
+
+        Parameters
+        ----------
+        """
+        # Compute all estimators
+        # For the moment only consider estimators compute per block.
+        # TODO: generalize for different block groups (loop over groups)
+        offset = self.num_walker_props
+        for k, e in self.items():
+            e.compute_estimator(walker_batch, hamiltonian, trial)
+            start = offset + self.get_offset(k)
+            end = start + int(self[k].size)
+            self.local_estimates[start:end] += e.data
+
+    def print_block(self, comm, block, walker_factors, div_factor=None):
+        self.local_estimates[: walker_factors.size] = walker_factors.buffer
+        comm.Reduce(self.local_estimates, self.global_estimates, op=MPI.SUM)
+        output_string = " "
+        # Get walker data.
+        offset = walker_factors.size
+        if comm.rank == 0:
+            walker_factors.post_reduce_hook(self.global_estimates[:offset], block)
+        output_string += walker_factors.to_text(self.global_estimates[:offset])
+        output_string += " "
+        for k, e in self.items():
+            if comm.rank == 0:
+                start = offset + self.get_offset(k)
+                end = start + int(self[k].size)
+                est_data = self.global_estimates[start:end]
+                e.post_reduce_hook(est_data)
+                est_string = e.data_to_text(est_data)
+                e.to_ascii_file(est_string)
+                if e.print_to_stdout:
+                    output_string += est_string
+        if comm.rank == 0:
+            shift = self.global_estimates[walker_factors.get_index("HybridEnergy")]
+
+        else:
+            shift = None
+        walker_factors.eshift = comm.bcast(shift)
+        if comm.rank == 0:
+            self.output.push_to_chunk(self.global_estimates, f"data")
+            self.output.increment()
+        if comm.rank == 0:
+            print(f"{block:>17d} " + output_string)
+        self.zero()
+    
+    def print_cut(self, comm, block, t, walker_factors):
```

Comment:
What does this function do?

Reply:
Prints the energy and particle number estimators at each time slice. It's used to generate the imaginary time trace of estimators

Reply:
Cool. Maybe rename it print_time_slice, and add a docstring.